### PR TITLE
Generic programming in c#

### LINKED ISSUE #2624: Understanding generic programming of C# in detail.
### Proposed title of article
Introduction to generic programming in C#.

### Introduction paragraph (2-3 paragraphs):
Due to the great advancement of technology in the 21st century, many software systems require to store large groups of entities
rather than individual entry. The concept of array could solve this problem but in C#, generic collections are far much better.

Collections according to C# are classes that hold together groups of objects. The .NET libraries provide useful system defined classes, like collection namespace, which  provide various collections and their functionalities.

In this article lists, dictionaries and hashsets collections are worked into and explained in detailed and easily understandable manner.

### Key takeaways: 
1. Understand the concept of list in C#.
2. Understand dictionaries in C#.
3. Understand hashsets form of collection in C#.
4. Know the advantages of a generic collection over a non-generic collection.

### References: 
N/A

### PR DESCRIPTION
**Pull request by elijahndungu**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
This fixes #2624

### CODE REVIEW

**Comment by john-amiscaray** [comment]
@hectorkambow I would like to review this article please

**Comment by hectorkambow** [comment]
@john-amiscaray  will be doing a test review on this PRs. 

cc @louisefindlay23 @ahmadmardeni1 @LinusMuema can answer any questions that pop up - and changes the labels when needed.

**Comment by john-amiscaray** [comment]
@elijahndungu Hello there, welcome to the Engineering Education program. I will be your peer reviewer for your first article. I know for me my first article was a learning experience so don't worry if you make a bunch more mistakes your first time. After my first article, it was smooth sailing for me. Looking forward to seeing what value you have to bring to others in this article!

**Comment by john-amiscaray** [comment]
@elijahndungu I just finished going through your article and took note of some specific feedback I would like to give you. Overall, it was a pretty good first article, there weren't really any major errors just a few smaller ones. I do have a couple of general tips to share with you, however. First: make sure all of your headings use 3 *#*'s or more. There were some instances where you used two or one. Also, at one point you italicized something using underscores (*_*), I believe I remember that it is a convention here to use astericks (*\**) instead. Finally, there were some overly long sentences. When you are checking the article in [Hemmingway](https://hemingwayapp.com/), try to have 0 sentences highlighted in red. But overall, good job. I will be posting the specific, line-to-line suggestions/fixes some time today.

**Comment by ahmadmardeni1** [comment]
Can you please make comments on the lines you are talking about? You can find a small + sign on the right of every line in the article (index.md file). @john-amiscaray

**Comment by john-amiscaray** [comment]
> 
> 
> Can you please make comments on the lines you are talking about? You can find a small + sign on the left of every line in the article (index.md file). @john-amiscaray

@ahmadmardeni1  Yes, I will be doing that soon. I just had to go somewhere today and was reviewing a second article so I could not do that immediately. I meant to say that I would post those later today, sorry if that was not clear.

File: content/articles/generic-programming-in-c#/index.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+But due to the great advancement of technology in the 21st century, many software systems require to store large groups of entities
+rather than individual entry. The concept of non-generic data handling can't fully solve this problem therefore in C#, generic collections are designed to solve this problem. Generic is a class that gives users the ability to define classes and methods with the placeholder.
+
+### Collections
+**Collections** according to C# are classes that hold together groups of objects. The `.NET` libraries provide useful system-defined classes, like collection namespace, which provide various collections and their functionalities. The use of non-generic collections is normally discouraged therefore my article fully focuses on generic collections.
+
+We have three common generic collections incorporated in the C# namespace.
+
+## 1. LIST
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:16]: > ## 1. LIST

Headings must have 3 *#*'s or more as a convention for our articles. Also, I would put this heading to have 4 *#*'s to make it a subheading under *Collections* since a list is a type of collection. Also, it's not necessary to put it in all caps and it should be plural. So overall change this line to:

> \#\#\#\# 1. List

File: content/articles/generic-programming-in-c#/index.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+            gen.GenFunction("Hello world");
+            
+        }
+    }
+}
+```
+
+Note that _<T>_ is a placeholder for data types and replaced by datatype at compile time.
+
+### 2. Hashsets
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:86]: > \#\#\# 2. Hashsets

Like I said with the *lists* heading, I would have this as four *#*'s, to make this a subheading under the *Collections* section.

File: content/articles/generic-programming-in-c#/index.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+A list is the most commonly used generic collection. Its because they are much more convenient when they are used instead of an array.
+A list is a very general data structure with each entry occupying a specific position known as an index. Lists are very flexible as they automatically resize every time data is added.
+
+Lists are very easy to manipulate. This is because many methods have been created and abstracted, meaning the bulky work is already done for the user. The user only has to call the method and manipulate the list.
+ 
+A list stores data in sequence, although the sequence oughtn't to have to be particular. It can also accept duplicate entries.
+Below is a pictorial explanation of a list. A generic list class is declared as shown:
+
+```
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:26]: > ```
> public class genericList<T>
> {
>     
>     void add(T input){}
> }
> ```

This code snippet is missing syntax highlighting add *csharp* after the upper 3 backticks (*`*) like so:

> \`\`\`csharp
> // your code goes here
> \`\`\`

so your code has syntax highlighting like so:

```csharp
using System;
using System.Collections.Generic;  
  
class MyDictinary {
    static public void Main () {
          
      
        //  Creating a dictionary using Dictionary <TKey,TValue> class
        Dictionary<int, string> My_dict1 = new Dictionary<int, string>(); 
            
          // Adding <key/value> pair in the Dictionary using Add() method
          My_dict1.Add(1123, "Welcome");
          My_dict1.Add(1124, "Hello");
          My_dict1.Add(1125, "World");
            
          foreach(KeyValuePair<int, string> ele1 in My_dict1)
          {
              Console.WriteLine("{0} and {1}", ele1.Key, ele1.Value);
          }
          Console.WriteLine();
            
          // Creating another dictionary
 
      Dictionary<string, string> My_dict2 = new Dictionary<string, string>(){
                                  {"a.1", "CSharp"},
                                  {"a.2", "Java"},
                                {"a.3", "Python"} }; 
           
          foreach(KeyValuePair<string, string> ele2 in My_dict2)
          {
              Console.WriteLine("{0} and {1}", ele2.Key, ele2.Value);
          }
    }
}
```

File: content/articles/generic-programming-in-c#/index.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+public class genericList<T>
+
+{
+    
+    void add(T input){}
+}
+```
+
+On the same note, a list can be of integer or string data type. They are declared differently but the syntax is the same as shown:
+```
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:36]: > ```
> using system;
> {
> private class genericListExample
> static void Main ()
> {
>     genericlist<int> list1 = new genericlist<int>(); //This declares a  list of type int.
>     
>      genericlist<string> list2 = new genericlist<string>(); //This declares a list of type string.
>      
>       genericlist<int> list1 = new genericlist<int>(); //This declares a list of type int.
> }
> }
> ```

As I said before, this is missing syntax highlighting, see the comment on your first code example (the `genericList` class) for details.

File: content/articles/generic-programming-in-c#/index.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+     genericlist<string> list2 = new genericlist<string>(); //This declares a list of type string.
+     
+      genericlist<int> list1 = new genericlist<int>(); //This declares a list of type int.
+}
+}
+```
+
+A simple c# example of generic list implementation might help in understanding.
+
+```
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:54]: > ```
> using system;
> using system.collection.generic; // A system namespace for implementing generic lists.
> namespace GenericListExample
> {
>     public class Generic<T>  // Declare Generics.
>     {
>         public void GenFunction(T printvalue)
>         {
>             Console.WriteLine(printvalue)
>         }
>     }
>     public class program
>     {
>         public static void Main(string [] args)
>         {
>             Console.WriteLine("Printing integer value");
>             Generic<int> gen = new Generic<int>();
>             gen.GenFunction(10);
>             
>             Console.WriteLine("Printing string value");
>             Generic<string> genString = new Generic<string>();
>             gen.GenFunction("Hello world");
>             
>         }
>     }
> }
> ```

This is missing syntax highlighting, see the comment on your first code example (the `genericList` class) for details.

File: content/articles/generic-programming-in-c#/index.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+            Console.WriteLine("Printing string value");
+            Generic<string> genString = new Generic<string>();
+            gen.GenFunction("Hello world");
+            
+        }
+    }
+}
+```
+
+Note that _<T>_ is a placeholder for data types and replaced by datatype at compile time.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:84]: > Note that _<T>_ is a placeholder for data types and replaced by data type at compile time

There is a bit of a formatting issue here. In markdown, *\>* indicates that the following paragraph is a blockquote. You need to escape the *\>* by placing a *\\* in front like so: 

> _<T\\>_ 

This way it look like this:

> _<T\>_

instead of:

> _<T>_

(doesn't even appear). 

Also: replace the *_* 's with *\**'s

Comment (john-amiscaray) [reply content/articles/generic-programming-in-c#/index.md:84]: > Note that _<T>_ is a placeholder for data types and replaced by datatype at compile time.

Change the wording to:

> Note that *<T\>* is a placeholder that is replaced by a data type at compile time. 

To be more grammatically correct and concise.

File: content/articles/generic-programming-in-c#/index.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
+
+### Introduction
+
+In any modern system, data handling is one of the major and vital goals. The way data is handled is what makes the total difference. Some systems go for generic one which includes lists, dictionaries, and hash.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:4]: > Some systems go for generic one which includes lists, dictionaries, and hash.

Change this to:

> Some systems go for a generic approach which includes the use of lists, dictionaries, and hash.

To be more grammatically correct and specific. Also, I was a little confused about the meaning of *hash* in this context because that word can have [many meanings](https://cs.stackexchange.com/a/55472). So you may need to clarify or reword it a bit here.

Comment (john-amiscaray) [reply content/articles/generic-programming-in-c#/index.md:4]: > Some systems go for generic one which includes lists, dictionaries, and hash.
> 
> Others prefer non-generic which are arrays, array lists, and bit arrays.

These two sentences are closely related to each other and should be in the same paragraph.

File: content/articles/generic-programming-in-c#/index.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+Others prefer non-generic which are arrays, array lists, and bit arrays. Non- generic programming represents an ordered collection of an object that can be indexed individually. As a matter of fact, generic programming is strongly typed while non- generic programming is not strongly typed, which means generic programming enforces strict rules on blending of values with different data types.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:6]: > As a matter of fact, generic programming is strongly typed while non- generic programming is not strongly typed, which means generic programming enforces strict rules on blending of values with different data types.

*As a matter of fact* may not be necessary here and is a bit of a weird way to transition into the idea of strongly/weakly typed programming. That phrase is usually used to introduce proof of something said previously. Also, you remove the space between *non-* and *generic*, i.e. change:

> non- generic

to:

> non-generic

File: content/articles/generic-programming-in-c#/index.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+But due to the great advancement of technology in the 21st century, many software systems require to store large groups of entities
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:8]: > But due to the great advancement of technology in the 21st century, many software systems require to store large groups of entities
rather than individual entry. 

This sentence is a little long, consider shortening it or splitting it up. Also, you should pluralize *individual entry* (i.e. change it to *individual entries*).

File: content/articles/generic-programming-in-c#/index.md [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+rather than individual entry. The concept of non-generic data handling can't fully solve this problem therefore in C#, generic collections are designed to solve this problem. Generic is a class that gives users the ability to define classes and methods with the placeholder.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:9]: > The concept of non-generic data handling can't fully solve this problem therefore in C#, generic collections are designed to solve this problem

I would probably split: 
> therefore in C#, generic collections are designed to solve this problem

into another sentence since it is a new idea. Maybe say something like:

> To solve this problem, C# introduced generic collections. 

Or you could just put it into a new sentence as is, that would be fine too.

File: content/articles/generic-programming-in-c#/index.md [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+
+### Collections
+**Collections** according to C# are classes that hold together groups of objects. The `.NET` libraries provide useful system-defined classes, like collection namespace, which provide various collections and their functionalities. The use of non-generic collections is normally discouraged therefore my article fully focuses on generic collections.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:12]: > like collection namespace

Add a *the* in front of *collection namespace*.

File: content/articles/generic-programming-in-c#/index.md [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+We have three common generic collections incorporated in the C# namespace.
+
+## 1. LIST
+
+A list is the most commonly used generic collection. Its because they are much more convenient when they are used instead of an array.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:18]: > Its because they are much more convenient when they are used instead of an array.

Change: *Its* to *It’s* to be more grammatically correct. Also, consider changing:

> much more convenient when they are used instead of an array

To something like:

> a very convenient substitute for an array

And maybe use the word *advantageous* instead of convenient but that’s up to you. This is just to be more concise.

File: content/articles/generic-programming-in-c#/index.md [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+A list is a very general data structure with each entry occupying a specific position known as an index. Lists are very flexible as they automatically resize every time data is added.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:19]: > A list is a very general data structure with each entry occupying a specific position known as an index. Lists are very flexible as they automatically resize every time data is added.
> 
> Lists are very easy to manipulate.

A lot of repetition of the word *list* which sounds choppy. I would change the second sentence to say *They* instead of *Lists*.

File: content/articles/generic-programming-in-c#/index.md [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+
+Lists are very easy to manipulate. This is because many methods have been created and abstracted, meaning the bulky work is already done for the user. The user only has to call the method and manipulate the list.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:21]: > This is because many methods have been created and abstracted, meaning the bulky work is already done for the user. The user only has to call the method and manipulate the list.

This could be more concise and the grammar in the second sentence is a little off. I would say something like:

> They contain many powerful methods that abstract complex work from the user. Manipulating a list is as easy as calling these methods.

File: content/articles/generic-programming-in-c#/index.md [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+
+On the same note, a list can be of integer or string data type. They are declared differently but the syntax is the same as shown:
+```
+using system;
+
+{
+private class genericListExample
+static void Main ()
+{
+    genericlist<int> list1 = new genericlist<int>(); //This declares a  list of type int.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:43]: > //This declares a  list of type int.

Remove the second space after *a* and before *list*. Also, I personally would add a space between the *//* and *This* but as long as you are consistent it should be fine.

File: content/articles/generic-programming-in-c#/index.md [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+            
+        }
+    }
+}
+```
+
+Note that _<T>_ is a placeholder for data types and replaced by datatype at compile time.
+
+### 2. Hashsets
+This is an implementation of a set. A set is like a big bag of entries with no particular order rather than a list. They are based on the discrete mathematic's idea of a 'set', which is the grouping of entries that can contain one, two, or many distinct items.  Sorting in a hash set is impossible because the entries have to be sorted. Use of a hash set is very much applicable where we have to store a large amount of data that doesn't have to be sorted eg storing the identity of cars passing through an entrance gate. A program example:
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:87]: > rather than a list

I would probably rephrase it to say *unlike a list*. It just feels like a more appropriate word choice. Although it shouldn’t make too large of a difference.

Comment (john-amiscaray) [reply content/articles/generic-programming-in-c#/index.md:87]: > Sorting in a hash set is impossible because the entries have to be sorted

It’s hard to understand what you are trying to say here. I would revise this. First, you say that sorting a hash set is impossible, but then follow up by saying it must be sorted. That second part sounds like you are saying it’s possible to sort it which is contradictory.

Comment (john-amiscaray) [reply content/articles/generic-programming-in-c#/index.md:87]: > items.  Sorting

There seems to be an extra space here after the period

Comment (john-amiscaray) [reply content/articles/generic-programming-in-c#/index.md:87]: > Use of a hash set is very much applicable where we have to store a large amount of data that doesn't have to be sorted eg storing the identity of cars passing through an entrance gate

I would probably separate the example (the part after the *eg*) into its own sentence. That sentence is rather long. You could at very least put a comma between the first part and the example, but I’d highly recommend making the example its own sentence.

Comment (john-amiscaray) [reply content/articles/generic-programming-in-c#/index.md:87]: > A program example:

I would probably rephrase this to something like:

> Here is an example program to help you understand this concept:

Since *example* is acting as an adjective to *program* which is a noun. Also just saying, *an example program* doesn’t flow well, so I altered it further to connect more with what was being said.

File: content/articles/generic-programming-in-c#/index.md [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+
+```
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:89]: > ```
>  using System;
> using System.Collections.Generic;
>   
> class MyHashSet {
>     static public void Main()
>     {
>   
>         // Creating HashSet
>      
>         HashSet<string> myhash1 = new HashSet<string>();
>   
>         // Add the elements in HashSet
>   
>         myhash1.Add("Ndung'u");
>         myhash1.Add("Elijah");
>         // Creating another HashSet
>      
>         HashSet<string> myhash2 = new HashSet<string>();
>         myhash2.Add("Section");
>         myhash2.Add("Git");
>    
>   
>         // Using ExceptWith method
>         myhash1.ExceptWith(myhash2);
>         foreach(var ele in myhash1)
>         {
>             Console.WriteLine(ele);
>         }
>     }
> }
> ```

This is missing syntax highlighting, see the comment on your first code example (the genericList class) for details.

File: content/articles/generic-programming-in-c#/index.md [hunk 18]
```diff
@@ -0,0 +1,182 @@
...
+        {
+            Console.WriteLine(ele);
+        }
+    }
+}
+```
+
+We note that many operations available for the list are available for hash sets except that:
+
+* We CANNOT add an object at a specific position since the elements are not in any order.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:126]: > We CANNOT add an object at a specific position since the elements are not in any order.

I would say it’s not necessary to put *CANNOT* in all caps. If you really want to emphasize it I would bold or italicize it.

File: content/articles/generic-programming-in-c#/index.md [hunk 19]
```diff
@@ -0,0 +1,182 @@
...
+* We cannot replace an item for the same reason, although we can add and delete another.
+* Retrieving all the items is possible but the sequence is indeterminate.
+* It's meaningless to find what position an element is or rather to sort.
+
+### 3. Dictionaries
+Dictionaries are slightly different from lists and sets because instead of individual objects, they store pairs; a key and a value.
+The key contains the address to the pair. The value is the piece of information associated with the key.
+Dictionaries have a very wide market application for instance an address book in which keys are people's names, and the values are address, phone, email, etc.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:134]: > Dictionaries have a very wide market application for instance an address book in which keys are people's names, and the values are address, phone, email, etc.

I recommend separating the example into its own sentence.

File: content/articles/generic-programming-in-c#/index.md [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+Duplicate keys are not permitted though values are. Also, a dictionary doesn't store items in order, but it can be sorted in order of keys to store a sorted dictionary. A simple program to illustrate:
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:135]: > A simple program to illustrate:

I would revise this to:

> Here is a simple program to illustrate:


Like I said before the other time you were introducing an example program, the transition to it wasn’t smooth. So I rephrased it to make it connect more with what was said last.

File: content/articles/generic-programming-in-c#/index.md [hunk 21]
```diff
@@ -0,0 +1,182 @@
...
+
+```
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:137]: > ```
> using System;
> using System.Collections.Generic;  
>   
> class MyDictinary {
>     static public void Main () {
>           
>       
>         //  Creating a dictionary using Dictionary <TKey,TValue> class
>         Dictionary<int, string> My_dict1 = new Dictionary<int, string>(); 
>             
>           // Adding <key/value> pair in the Dictionary using Add() method
>           My_dict1.Add(1123, "Welcome");
>           My_dict1.Add(1124, "Hello");
>           My_dict1.Add(1125, "World");
>             
>           foreach(KeyValuePair<int, string> ele1 in My_dict1)
>           {
>               Console.WriteLine("{0} and {1}", ele1.Key, ele1.Value);
>           }
>           Console.WriteLine();
>             
>           // Creating another dictionary
>  
>       Dictionary<string, string> My_dict2 = new Dictionary<string, string>(){
>                                   {"a.1", "CSharp"},
>                                   {"a.2", "Java"},
>                                 {"a.3", "Python"} }; 
>            
>           foreach(KeyValuePair<string, string> ele2 in My_dict2)
>           {
>               Console.WriteLine("{0} and {1}", ele2.Key, ele2.Value);
>           }
>     }
> }
> ```

This is missing syntax highlighting, see the comment on your first code example (the `genericList` class) for details.

File: content/articles/generic-programming-in-c#/index.md [hunk 22]
```diff
@@ -0,0 +1,182 @@
...
+
+We note that many operations available for the list are available for hash sets except that:
+
+* We CANNOT add an object at a specific position since the elements are not in any order.
+* We cannot replace an item for the same reason, although we can add and delete another.
+* Retrieving all the items is possible but the sequence is indeterminate.
+* It's meaningless to find what position an element is or rather to sort.
+
+### 3. Dictionaries
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:131]: > \#\#\# 3. Dictionaries

Change this to:

> \#\#\#\# 3. Dictionaries

Like I said with the headers for the other data structures, I would make this a subheading of the *collections* section. To do this, put \#\#\#\# to denote the header.

File: content/articles/generic-programming-in-c#/index.md [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+           
+          foreach(KeyValuePair<string, string> ele2 in My_dict2)
+          {
+              Console.WriteLine("{0} and {1}", ele2.Key, ele2.Value);
+          }
+    }
+}
+```
+
+## Conclusion
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:174]: > \#\# Conclusion

Change this to:

> \#\#\# Conclusion

Headings must have at least 3 *#*'s

File: content/articles/generic-programming-in-c#/index.md [hunk 24]
```diff
@@ -0,0 +1,182 @@
...
+In summary, I have explained what are generics in c#, its advantages, and a complete programming example
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:175]: > In summary, I have explained what are generics in c#, its advantages, and a complete programming example   
> I also explained the comparison between generic and non - generic programming where both can add, delete, populate and sort Generic List. Lists, dictionaries, and hash sets are automatically resized every time an item is added or deleted.

The first sentence is missing a period in the end. Also in the second sentence, change: *non - generic* to *non-generic*; there shouldn’t be spaces before and after the *-*. Also, there are a couple of grammatical errors that need to be fixed. Also, personally, I would prefer to say: *we went through* or *we learned* instead of *I explained*. The tone just sounds more like you are trying to help and educate someone and sounds friendlier in a way. To fix all these, I would say something like:

File: content/articles/generic-programming-in-c#/index.md [hunk 25]
```diff
@@ -0,0 +1,182 @@
...
+          {
+              Console.WriteLine("{0} and {1}", ele2.Key, ele2.Value);
+          }
+    }
+}
+```
+
+## Conclusion
+In summary, I have explained what are generics in c#, its advantages, and a complete programming example   
+I also explained the comparison between generic and non - generic programming where both can add, delete, populate and sort Generic List. Lists, dictionaries, and hash sets are automatically resized every time an item is added or deleted.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:176]: > In summary, I have explained what are generics in c#, its advantages, and a complete programming example   
> I also explained the comparison between generic and non - generic programming where both can add, delete, populate and sort Generic List. Lists, dictionaries, and hash sets are automatically resized every time an item is added or deleted.

The ending sounds a little abrupt. It is mentioned in the new contributor’s guide that you should end off by urging the reader to use the knowledge you gave them, and suggest ways to do so.

File: content/articles/generic-programming-in-c#/index.md [hunk 26]
```diff
@@ -0,0 +1,182 @@
...
+    genericlist<int> list1 = new genericlist<int>(); //This declares a  list of type int.
+    
+     genericlist<string> list2 = new genericlist<string>(); //This declares a list of type string.
+     
+      genericlist<int> list1 = new genericlist<int>(); //This declares a list of type int.
+}
+}
+```
+
+A simple c# example of generic list implementation might help in understanding.
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:52]: > A simple c# example of generic list implementation might help in understanding.

The grammar here is a bit off. I would change this to:

> A simple implementation of a generic list in c# might help you understand:

**Comment by john-amiscaray** [comment]
@elijahndungu I added the specific, in-line comments on your article so you can start implementing my suggestions. I also forgot to mention in the general tips I gave you that your code snippets are missing syntax highlighting. Whenever you put a code sample, make sure to add the name of the language after the first three backticks like so:

> \`\`\`csharp
// your code goes here
> \`\`\`

so you have syntax highlighting on your code like this:

```csharp
public class genericList<T>

{
    
    void add(T input){}
}
```

File: content/articles/generic-programming-in-c#/index.md [hunk 27]
```diff
@@ -0,0 +1,182 @@
+
```
Comment (john-amiscaray) [inline content/articles/generic-programming-in-c#/index.md:1]: When I submit my articles, I normally like adding the title and a description at the top. Consider doing this, it is pretty helpful.

**Comment by ninjaginja** [comment]
plagiarism check completed - all good
[2120420-7f1e5701c363b403a71c-1626807875715.pdf](https://github.com/section-engineering-education/engineering-education/files/6850682/2120420-7f1e5701c363b403a71c-1626807875715.pdf)

**Comment by elijahndungu** [comment]
@john-amiscary I have made the changes, you can look at them. Thank you

**Comment by hectorkambow** [comment]
@WanjaMIKE  Just wanted to get a status update on this article whenever you get a chance. 

This PR has been in our queue for quite a bit. 

Thank you

**Comment by WanjaMIKE** [comment]
Working on it @hectorkambow

**Review by WanjaMIKE — APPROVED** [review]
Please go through the changes I have made, as well as the comments from the previous reviewer to improve further in your future article. Also, ensure that all your code snippets work. I corrected numerous bugs in the code snippets.

**Comment by hectorkambow** [comment]
Overall great work Gents @elijahndungu 

cc @WanjaMIKE 🚀
